### PR TITLE
Export file and refactor random utils

### PR DESCRIPTION
**Pull request by catsuryuu**

### CODE REVIEW

File: src/ui/hexedit.cc [hunk 1]
```diff
@@ -1341,24 +1346,49 @@ void HexEdit::saveChunkToFile(const QString& path) {
   saveDataToFile(byte_offset, size, path);
 }
 
+/**
+ * The data is saved to temporary file and then moved to destined location,
```
Comment (mkow) [inline src/ui/hexedit.cc]: `a temporary file`

Comment (mkow) [reply src/ui/hexedit.cc]: `to destined location` -> `to the destination`

File: src/ui/hexedit.cc [hunk 2]
```diff
@@ -1341,24 +1346,49 @@ void HexEdit::saveChunkToFile(const QString& path) {
...
+ * so that the previous content wouldn't be destroyed if something went wrong
```
Comment (mkow) [inline src/ui/hexedit.cc]: `content` -> `contents`
`went` -> `go`

File: src/ui/hexedit.cc [hunk 3]
```diff
@@ -1341,24 +1346,49 @@ void HexEdit::saveChunkToFile(const QString& path) {
...
+ */
 void HexEdit::saveDataToFile(qint64 byte_offset, qint64 size,
                              const QString& path) {
   if (byte_offset + size > dataBytesCount_) {
     size = dataBytesCount_ - byte_offset;
   }
+  auto data = edit_engine_.bytesValues(byte_offset, size);
 
-  auto data_to_save = edit_engine_.bytesValues(byte_offset, size);
+  QString tmp_path = path + ".tmp." + util::generateRandomUppercaseText(8);
```
Comment (mkow) [inline src/ui/hexedit.cc]: 12 chars would be safer

File: src/ui/hexedit.cc [hunk 4]
```diff
@@ -1341,24 +1346,49 @@ void HexEdit::saveChunkToFile(const QString& path) {
...
 
-  QFile file(path);
+  // It will be closed automatically.
```
Comment (mkow) [inline src/ui/hexedit.cc]: No need to say this, this is a general convention used everywhere in C++.

File: src/ui/hexedit.cc [hunk 5]
```diff
@@ -1341,24 +1346,49 @@ void HexEdit::saveChunkToFile(const QString& path) {
...
+  QFile file(tmp_path);
   if (!file.open(QIODevice::WriteOnly)) {
-    QMessageBox::information(this, tr("Unable to open file"),
-                             file.errorString());
+    QMessageBox::warning(this, tr("Unable to create tmp file"),
+                         tr("Unable to create tmp file '%1': %2")
```
Comment (mkow) [inline src/ui/hexedit.cc]: `tmp` -> `temporary` in the longer explanation

File: src/ui/hexedit.cc [hunk 6]
```diff
@@ -1341,24 +1346,49 @@ void HexEdit::saveChunkToFile(const QString& path) {
...
+    if (!QFile::remove(path)) {
+      QMessageBox::warning(this, tr("Error"),
+                           tr("File \'%1\' cannot be replaced.").arg(path));
+      return;
+    }
+  }
+  if (!file.rename(path)) {
+    QMessageBox::warning(
+        this, tr("Error"),
+        tr("Error moving tmp file '%1' to '%2'.").arg(tmp_path).arg(path));
```
Comment (mkow) [inline src/ui/hexedit.cc]: ditto

File: include/ui/hexeditwidget.h [hunk 7]
```diff
@@ -98,17 +96,17 @@ class HexEditWidget : public View {
...
 
   QToolBar* file_tool_bar_;
   QToolBar* edit_tool_bar_;
   QToolBar* tools_tool_bar_;
   QToolBar* view_tool_bar_;
 
-  QAction* upload_act_;
   QAction* save_as_act_;
-  QAction* save_readable_;
+  //  QAction* save_readable_;
```
[the remark this answers is not in the document]
Comment (catsuryuu) [reply include/ui/hexeditwidget.h]: Done

File: include/util/random.h [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+#pragma once
+
+#include <algorithm>
+#include <array>
+#include <functional>
+#include <random>
```
[the remark this answers is not in the document]
Comment (catsuryuu) [reply include/util/random.h]: Done

File: include/util/random.h [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+#include <QString>
+
+namespace veles {
+namespace util {
+
+extern std::mt19937 global_mersenne_twister;
```
[the remark this answers is not in the document]
Comment (catsuryuu) [reply include/util/random.h]: Done

File: include/util/random.h [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+
+QString generateRandomUppercaseText(qint64 size);
```
[the remark this answers is not in the document]
Comment (catsuryuu) [reply include/util/random.h]: Done

File: src/ui/hexedit.cc (hunk 1 above)
Comment (catsuryuu) [reply src/ui/hexedit.cc]: Done

File: src/ui/hexedit.cc (hunk 3 above)
Comment (catsuryuu) [reply src/ui/hexedit.cc]: Done

File: src/ui/hexedit.cc (hunk 4 above)
Comment (catsuryuu) [reply src/ui/hexedit.cc]: Removed

File: src/ui/hexedit.cc (hunk 5 above)
Comment (catsuryuu) [reply src/ui/hexedit.cc]: Done

File: src/ui/hexedit.cc (hunk 6 above)
Comment (catsuryuu) [reply src/ui/hexedit.cc]: Done

File: src/ui/hexedit.cc [hunk 11]
```diff
@@ -1341,24 +1346,49 @@ void HexEdit::saveChunkToFile(const QString& path) {
...
     return;
   }
 
-  QDataStream stream(&file);
-  stream.writeRawData(reinterpret_cast<const char*>(data_to_save.rawData()),
-                      static_cast<int>(data_to_save.octets()));
+  if (file.write(reinterpret_cast<const char*>(data.rawData()),
+                 static_cast<int>(data.octets())) == -1) {
+    QMessageBox::warning(this, tr("Error writing to tmp file"),
+                         tr("Error writing to tmp file '%1'.").arg(tmp_path));
```
[the remark this answers is not in the document]
Comment (catsuryuu) [reply src/ui/hexedit.cc]: Done

File: src/ui/hexedit.cc [hunk 12]
```diff
@@ -1341,24 +1346,49 @@ void HexEdit::saveChunkToFile(const QString& path) {
...
+    return;
+  }
+  if (QFile::exists(path)) {
+    if (!QFile::remove(path)) {
+      QMessageBox::warning(this, tr("Error"),
```
[the remark this answers is not in the document]
Comment (catsuryuu) [reply src/ui/hexedit.cc]: Done

File: src/ui/hexedit.cc [hunk 13]
```diff
@@ -1341,24 +1346,61 @@ void HexEdit::saveChunkToFile(const QString& path) {
...
+  }
+  if (!file.rename(path)) {
+    QMessageBox::warning(this, tr("Error"),
+                         tr("Error moving temporary file '%1' to '%2'.")
+                             .arg(tmp_path)
+                             .arg(path));
+    remove_tmp_file();
+    return;
+  }
 }
```
Comment (mkow) [inline src/ui/hexedit.cc]: `QFile` destructor can fail (when flushing buffers), we have to close it manually and check error code.

Comment (catsuryuu) [reply src/ui/hexedit.cc]: But `file.remove()` and `file.rename(path)` always close the file.

Comment (mkow) [reply src/ui/hexedit.cc]: I'm not sure if `QFile::rename` reopens the file after renaming, we should check it.

Comment (mkow) [reply src/ui/hexedit.cc]: The documentation is unclear, so I checked the source and it seems that `QFile::rename` always leaves the object in closed state.

File: include/util/random.h [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ */
+#pragma once
+
+#include <QString>
+
+namespace veles {
+namespace util {
+
+extern std::mt19937 g_mersenne_twister;
```
Comment (mkow) [inline include/util/random.h]: You use `std::mt19937` but not include it. It doesn't even build :P

File: src/ui/hexedit.cc (hunk 2 above)
Comment (mkow) [reply src/ui/hexedit.cc]: yup, my bad

**Pull request merged**